File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/AutoBlueRight.java
package org.firstinspires.ftc.teamcode;

import com.acmerobotics.roadrunner.Pose2d;
import com.acmerobotics.roadrunner.Vector2d;
import com.acmerobotics.roadrunner.ftc.Actions;
import com.qualcomm.robotcore.eventloop.opmode.Autonomous;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DistanceSensor;
import com.qualcomm.robotcore.hardware.IMU;
import com.qualcomm.robotcore.hardware.Servo;

@Autonomous(name = "#ClayAutoBlueRight", group = "Auto")
public final class AutoBlueRight extends LinearOpMode {
    private DcMotorEx hang, intake, slideR, slideL;
    private Servo trussL, trussR, armR, armL, clawR, clawL, drone, inlift;

    public int placementZone = 2;
    private IMU imu;

    private DistanceSensor distance;
    int liftPos = 0;
    public static void wait(int ms) {
        try {
            Thread.sleep(ms);
        } catch(InterruptedException ex) {
            Thread.currentThread().interrupt();
        }
    }


    @Override
    public void runOpMode() throws InterruptedException {

        intake = hardwareMap.get(DcMotorEx.class, "intake");
        slideR = hardwareMap.get(DcMotorEx.class, "slideR");
        slideL = hardwareMap.get(DcMotorEx.class, "slideL");
        drone = hardwareMap.get(Servo.class, "drone");
        trussR = hardwareMap.get(Servo.class, "trussR");
        trussL = hardwareMap.get(Servo.class, "trussL");
        armR = hardwareMap.get(Servo.class, "armR");
        armL = hardwareMap.get(Servo.class, "armL");
        clawR = hardwareMap.get(Servo.class, "clawR");
        clawL = hardwareMap.get(Servo.class, "clawL");
        inlift = hardwareMap.get(Servo.class, "inlift");
        distance = hardwareMap.get(DistanceSensor.class, "distance");
        imu = hardwareMap.get(IMU.class, "imu");
        clawR.setPosition(0.07);
        clawL.setPosition(0.29);

        inlift.setPosition(0.34);
        armR.setPosition(0.05);
        armL.setPosition(0.95);
        slideR.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        slideR.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        slideL.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        slideL.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        hang.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        Pose2d beginPose = new Pose2d(0, 0, 0);
            MecanumDrive drive = new MecanumDrive(hardwareMap, beginPose);
        waitForStart();
if (placementZone == 1) {
    Actions.runBlocking(
             drive.actionBuilder(beginPose)
                    .splineTo(new Vector2d(20.12, -6.98), Math.toRadians(-31.68))
                  .splineToConstantHeading(new Vector2d(7, 3), Math.toRadians(31.68))
                 .turnTo(Math.toRadians(0))
               .splineToConstantHeading(new Vector2d(49.28, 4.45), Math.toRadians(0))
                 .turnTo(Math.toRadians(90))
                     .splineToConstantHeading(new Vector2d(50.11, 76.62), Math.toRadians(90))
                     .turnTo(Math.toRadians(270))
                     .splineToConstantHeading(new Vector2d(19.38, 87.4), Math.toRadians(270))
                     .build()
    );

} else if (placementZone == 2) {
    Actions.runBlocking(
            drive.actionBuilder(beginPose)
                    .splineTo(new Vector2d(26.32, 2.04), 0)
                    .splineToConstantHeading(new Vector2d(16.8, -7.89), 0)
                    .splineToConstantHeading(new Vector2d(48.94, -8.83), 0)

                    .turnTo(Math.toRadians(90))
                    .splineToConstantHeading(new Vector2d(52.55, 77.19), Math.toRadians(90))
                    .turnTo(Math.toRadians(270))
                    .splineToConstantHeading(new Vector2d(28.21, 89.29), Math.toRadians(90))

                    .build()
    );
} else if (placementZone ==3){

                Actions.runBlocking(
                        drive.actionBuilder(beginPose)
                                .splineTo(new Vector2d(21.83, -5.68), Math.toRadians(-25))
                               .splineToConstantHeading(new Vector2d(25.96, 4.070), Math.toRadians(157.81))
                                .turnTo(Math.toRadians(270))
                                .splineToConstantHeading(new Vector2d(31.3, 31), Math.toRadians(270))
                                .build()
                );
        }

        liftPos=600;
        slideR.setTargetPosition(-liftPos);
        slideL.setTargetPosition(liftPos);

        slideR.setMode(DcMotor.RunMode.RUN_TO_POSITION);
        slideL.setMode(DcMotor.RunMode.RUN_TO_POSITION);

        slideR.setPower(1);
        slideL.setPower(1);

        wait(300);
        armL.setPosition(0.65);
        armR.setPosition(0.35);
        wait(300);
        clawL.setPosition(0);
        clawR.setPosition(0.36);
        wait(300);

        Actions.runBlocking(
                drive.actionBuilder(drive.pose)
                        .lineToY(83)
                        .build()
        );
        armR.setPosition(0.05);
        armL.setPosition(0.95);
        wait(100);
        liftPos=0;
        slideR.setTargetPosition(-liftPos);
        slideL.setTargetPosition(liftPos);

        slideR.setMode(DcMotor.RunMode.RUN_TO_POSITION);
        slideL.setMode(DcMotor.RunMode.RUN_TO_POSITION);

        slideR.setPower(1);
        slideL.setPower(1);
        inlift.setPosition(0);
        wait(200);
       Actions.runBlocking(
                drive.actionBuilder(drive.pose)
                        .strafeToConstantHeading(new Vector2d(50, 84.72))
                        .splineToConstantHeading(new Vector2d(50, 101.92), Math.toRadians(270))
                        .build()
        );


    }
}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/AutoBlueRight.kt
package org.firstinspires.ftc.teamcode

import com.acmerobotics.roadrunner.Pose2d
import com.acmerobotics.roadrunner.Vector2d
import com.acmerobotics.roadrunner.ftc.runBlocking
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.IMU
import com.qualcomm.robotcore.hardware.Servo

@Autonomous(name = "# Clay Blue Right")
class AutoBlueRight : LinearOpMode() {
    protected lateinit var intake: DcMotorEx
    protected lateinit var slideR: DcMotorEx
    protected lateinit var slideL: DcMotorEx
    protected lateinit var trussL: Servo
    protected lateinit var trussR: Servo
    protected lateinit var armR: Servo
    protected lateinit var armL: Servo
    protected lateinit var clawR: Servo
    protected lateinit var clawL: Servo
    protected lateinit var drone: Servo
    protected lateinit var inlift: Servo
    protected lateinit var imu: IMU

    var placementZone: SpikeMark = SpikeMark.LEFT
    var teamAlliance = Alliance.BLUE
    var elementSpikeMark = SpikeMark.RIGHT

    protected lateinit var drive: MecanumDrive

    open val beginPose = Pose2d(0.0, 0.0, 0.0)
    protected var liftPos = 0

    override fun runOpMode() {
        drive = MecanumDrive(hardwareMap, beginPose)

        intake =    hardwareMap[DcMotorEx::class.java,      "intake"    ]
        slideR =    hardwareMap[DcMotorEx::class.java,      "slideR"    ]
        slideL =    hardwareMap[DcMotorEx::class.java,      "slideL"    ]
        drone =     hardwareMap[Servo::class.java,          "drone"     ]
        trussR =    hardwareMap[Servo::class.java,          "trussR"    ]
        trussL =    hardwareMap[Servo::class.java,          "trussL"    ]
        armR =      hardwareMap[Servo::class.java,          "armR"      ]
        armL =      hardwareMap[Servo::class.java,          "armL"      ]
        clawR =     hardwareMap[Servo::class.java,          "clawR"     ]
        clawL =     hardwareMap[Servo::class.java,          "clawL"     ]
        inlift =    hardwareMap[Servo::class.java,          "inlift"    ]
        imu =       hardwareMap[IMU::class.java,            "imu"       ]



        clawR.position = 0.07
        clawL.position = 0.29
        inlift.position = 0.34

        // arm
        armR.position = 0.05
        armL.position = 0.95

        // mode settings
        slideR.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        slideR.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
        slideL.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        slideL.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER

        val autoSub = AutoSubsystem(this)
        autoSub.setAlliance(teamAlliance)

        /*   for (i in 0..100) {
               Thread.sleep(20)
               elementSpikeMark = autoSub.elementDetection()
               telemetry.addData("getMaxDistance", autoSub.pipeline.getMaxDistance())
               if (isStopRequested){
                   return
               }
           }*/

        while (!isStarted && !isStopRequested){
            elementSpikeMark = autoSub.elementDetection()
            autoSub.setAlliance(teamAlliance)
            telemetry.addLine("Select Alliance (Gamepad1 X = Blue, Gamepad1 B = Red)")
            telemetry.addData("Current Alliance Selected", teamAlliance.toString())
            telemetry.addData("Spike mark", autoSub.spikeMark == SpikeMark.LEFT)
            telemetry.update()
        }

        if (autoSub.spikeMark == SpikeMark.LEFT){
            elementSpikeMark = SpikeMark.LEFT
        } else if (autoSub.spikeMark == SpikeMark.CENTER){
            elementSpikeMark = SpikeMark.CENTER
        } else {
            elementSpikeMark = SpikeMark.RIGHT
        }


        runBlocking(when (placementZone) {
            SpikeMark.RIGHT -> drive.actionBuilder(beginPose)
                .splineTo(Vector2d(16.61, 0.66), Math.toRadians(0.0))
                .splineToConstantHeading(Vector2d(3.72, -18.46), Math.toRadians(0.0))
                .splineToConstantHeading(Vector2d(50.21, -16.11), Math.toRadians(0.0))
                .turnTo(Math.toRadians(90.0))
                .splineToConstantHeading(Vector2d(49.55, 73.64), Math.toRadians(90.0))
                .turnTo(Math.toRadians(270.0))
                .splineToConstantHeading(Vector2d(30.44, 88.65), Math.toRadians(270.0))
                .build()
            SpikeMark.CENTER -> drive.actionBuilder(beginPose)
                .splineTo(Vector2d(18.39, -0.21), 0.0)
                .splineToConstantHeading(Vector2d(4.96, -20.0), 0.0)
                .splineToConstantHeading(Vector2d(49.75, -18.46), 0.0)
                .turnTo(Math.toRadians(90.0))
                .splineToConstantHeading(Vector2d(50.0, 74.94), Math.toRadians(90.0))
                .turnTo(Math.toRadians(270.0))
                .splineToConstantHeading(Vector2d(25.42, 90.69), Math.toRadians(270.0))
                .build()
            SpikeMark.LEFT -> drive.actionBuilder(beginPose)
                .splineTo(Vector2d(20.32, 9.60), Math.toRadians(44.51))
                .splineToConstantHeading(Vector2d(2.98, -7.4), Math.toRadians(44.51))
                .turnTo(Math.toRadians(0.0))
                .splineToConstantHeading(Vector2d(50.4, -10.67), 0.0)
                .turnTo(Math.toRadians(90.0))
                .splineToConstantHeading(Vector2d(49.41, 73.98), Math.toRadians(90.0))
                .turnTo(Math.toRadians(270.0))
                .splineToConstantHeading(Vector2d(19.88, 88.94), Math.toRadians(270.0))
                .build()
        })
        liftPos = 600
        slideR.targetPosition = -liftPos
        slideL.targetPosition = liftPos
        slideR.mode = DcMotor.RunMode.RUN_TO_POSITION
        slideL.mode = DcMotor.RunMode.RUN_TO_POSITION
        slideR.power = 1.0
        slideL.power = 1.0
        sleep(300)
        armL.position = 0.65
        armR.position = 0.35
        sleep(300)
        clawL.position = 0.0
        clawR.position = 0.36
        sleep(300)
        runBlocking(
            drive.actionBuilder(drive.pose)
                .lineToY(83.0)
                .build()
        )
        armR.position = 0.05
        armL.position = 0.95
        sleep(100)
        liftPos = 0
        slideR.targetPosition = -liftPos
        slideL.targetPosition = liftPos
        slideR.mode = DcMotor.RunMode.RUN_TO_POSITION
        slideL.mode = DcMotor.RunMode.RUN_TO_POSITION
        slideR.power = 1.0
        slideL.power = 1.0
        inlift.position = 0.0
        sleep(200)
        runBlocking(
            drive.actionBuilder(drive.pose)
                .strafeToConstantHeading(Vector2d(50.0, 84.72))
                .splineToConstantHeading(Vector2d(50.0, 101.92), Math.toRadians(270.0))
                .build()
        )
    }
    }
